File: solutions/javascript/SubstringwithConcatenationofAllWords/SubstringwithConcatenationofAllWords.js
/**
 * @param {string} s
 * @param {string[]} words
 * @return {number[]}
 */
var findSubstring = function(s, words) {
    var res = [];
    if (s == '' || words.length == 0) {
        return res;
    }
    var len = words[0].length;
    var map = new Map();
    for (var i = 0; i < words.length; i++) {
        map.set(words[i], map.has(words[i]) ? map.get(words[i]) + 1 : 1);
    }
    for (var i = 0; i <= s.length - len * words.length; i++) {
        var copy = new Map(map);
        for (var j = 0; j < words.length; j++) {
            var str = s.substr(i + j * len, len);
            if (copy.has(str)) {
                var count = copy.get(str);
                if (count == 1) {
                    copy.delete(str);
                } else {
                    copy.set(str, count - 1);
                }
                if (copy.size == 0) {
                    res.push(i);
                    break;
                }
            } else {
                break;
            }
        }
    }
    return res;
};

File: solutions/java/SubstringwithConcatenationofAllWords/SubstringwithConcatenationofAllWords.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> res = new ArrayList<>();
        if ("".equals(s) || words.length == 0) {
            return res;
        }
        int len = words[0].length();
        Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i < words.length; i++) {
            map.put(words[i], map.containsKey(words[i]) ? map.get(words[i]) + 1 : 1);
        }
        for (int i = 0; i <= s.length() - len * words.length; i++) {
            Map<String, Integer> copy = new HashMap<>(map);
            for (int j = 0; j < words.length; j++) {
                int tmp = i + j * len;
                String str = s.substring(tmp, tmp + len);
                if (copy.containsKey(str)) {
                    int count = copy.get(str);
                    if (count == 1) {
                        copy.remove(str);
                    } else {
                        copy.put(str, count - 1);
                    }
                    if (copy.isEmpty()) {
                        res.add(i);
                        break;
                    }
                } else {
                    break;
                }
            }
        }
        return res;
    }
}